src/groups/mqb/mqbs/mqbs_filebackedstorage.t.cpp

```diff
@@ -0,0 +1,2010 @@
+                bool                            useSameGuids = false,
+                int                             refCount     = 1)
+    {
+        // PRECONDITIONS
+        BSLS_ASSERT_OPT(guidHolder);
+
+        int guidCount = static_cast<int>(guidHolder->size());
+
+        if (useSameGuids) {
+            BSLS_ASSERT_OPT(guidCount == msgCount);
+        }
+        else {
+            guidCount += msgCount;
+            guidHolder->reserve(guidCount);
+
+            for (int i = 0; i < msgCount; i++) {
+                bmqt::MessageGUID guid;
+                mqbu::MessageGUIDUtil::generateGUID(&guid);
+                guidHolder->push_back(guid);
+            }
+        }
+
+        for (int i = 0; i < msgCount; i++) {
+            const bmqt::MessageGUID& guid = guidHolder->at(guidCount -
+                                                           msgCount + i);
+
+            const int data = i + dataOffset;
+
+            mqbi::StorageMessageAttributes attributes(
+                static_cast<bsls::Types::Uint64>(data),
+                refCount,
+                bmqp::MessagePropertiesInfo::makeNoSchema(),
+                bmqt::CompressionAlgorithmType::e_NONE);
+
+            const bsl::shared_ptr<bdlbb::Blob> appDataPtr(
+                new (*s_allocator_p)
+                    bdlbb::Blob(&d_bufferFactory, s_allocator_p),
+                s_allocator_p);
+
+            bdlbb::BlobUtil::append(&(*appDataPtr),
+                                    reinterpret_cast<const char*>(&data),
+                                    static_cast<int>(sizeof(int)));
+
+            mqbi::StorageResult::Enum rc = d_replicatedStorage_mp->put(
+                &attributes,
+                guid,
+                appDataPtr,
+                appDataPtr);
+
+            if (rc != mqbi::StorageResult::e_SUCCESS) {
+                return rc;  // RETURN
+            }
+        }
+
+        return mqbi::StorageResult::e_SUCCESS;
+    }
+
+    void insertDataStoreRecord(mqbs::DataStoreRecordHandle*    handle,
+                               const mqbs::DataStoreRecordKey& key,
+                               const mqbs::DataStoreRecord&    record)
+    {
+        // PRECONDITIONS
+        BSLS_ASSERT_OPT(handle && "'handle' must be specified");
+
+        // TYPES
+        typedef mqbs::DataStoreConfig::RecordIterator RecordIterator;
+        typedef bsl::pair<RecordIterator, bool>       InsertRc;
+
+        InsertRc insertRc = d_records.insert(bsl::make_pair(key, record));
+        BSLS_ASSERT_SAFE(insertRc.second);
+        RecordIterator& recordItRef = *reinterpret_cast<RecordIterator*>(
+            handle);
+        recordItRef = insertRc.first;
+    }
+
+  private:
+    // NOT IMPLEMENTED
+    BaseTester(const BaseTester&) BSLS_KEYWORD_DELETED;
+    BaseTester& operator=(const BaseTester&) BSLS_KEYWORD_DELETED;
+};
+
+// =============
+// struct Tester
+// =============
+struct Tester : public BaseTester {
+    MockDataStore d_dataStore;
```

Comment:
Actually, I think there is a bug here.

In `Tester`, there is a field `MockDataStore d_dataStore`.
We pass a pointer to this field to `d_replicatedStorage_mp.load(...)`. But `d_replicatedStorage_mp` is a field of the base class `BaseTester`.

So when we call a chain of destructors, we first destroy fields of the child `Tester` class, and only after this we destroy fields of the parent `BaseTester` class.

This means, that the object under the pointer to `d_dataStore` will be destroyed before the `d_replicatedStorage_mp` is destroyed.

And we also call `d_replicatedStorage_mp->removeAll(k_NULL_KEY);` in `~BaseTester()`, so we basically access the `d_dataStore` object which was already destroyed.

Reply:
Also, if we pass a pointer to a field value somewhere else, we should probably protect ourselves from incorrect use of this class, by removing copy/assignment operators

Reply:
Makes sense. I'll put all the code into the single Tester class